solr/core/src/java/org/apache/solr/cloud/api/collections/DeleteReplicaCmd.java

```diff
@@ -147,14 +170,27 @@ void deleteReplicaBasedOnCount(ClusterState clusterState,
       }
     }
 
+    if (placementPlugin != null) {
```

Comment:
By adding `if (placementPlugin != null)` logic in the `*Cmd` classes, we are breaking the encapsulation that placement logic is handled by `Assign.AssignStrategy`.
The only reason `*Cmd` code currently (before this PR) is even aware of the notion of `PlacementPlugin` is because `PlacementPluginAssignStrategy` configuration (rather than `LegacyAssignStrategy`) is dependent on a plugin being defined...
I suggest to move all `*Cmd` vetting logic to `Assign.AssignStrategy`, so that `*Cmd` only need to pass the instance of `PlacementPlugin` to that code (and later when we finally decide how clusters are configured this will go away, and we keep `*Cmd` clean of any specific assign strategy behavior).

In `Assign.AssignStrategy` the default vetting logic will be "accept all" and in `PlacementPluginAssignStrategy` we can implement the checks we like.

Reply:
In `Assign.AssignStrategy` if we want to be exhaustive, we should for example reject shard splits for secondary that are targets of `withCollection` (given we refuse such targets to have more than one shard).
Not saying we should do it, but the vetting infra we put in place should allow logical extension to all these aspects (with minor impact on the commands).

Also, pushing all the logic to `Assign.AssignStrategy` and minimizing changes to `*Cmd` limits the impact of a regression (moving to `LegacyAssignStrategy` should be a workaround for most problems).